lib/navigation/preprocess_keyence.cpp

```diff
@@ -4,28 +4,32 @@
 
 namespace hyped::navigation {
 
-KeyencePreprocessor::KeyencePreprocessor(hyped::core::ILogger &logger) : log_(logger)
+KeyencePreprocessor::KeyencePreprocessor(core::ILogger &logger)
+    : log_(logger),
+      previous_data_status_(KeyenceDataStatus::kAgreed)
 {
-  // Currently no extra parameters so constructor is empty
-  has_keyence_disagreed_ = false;
 }
-
 SensorChecks KeyencePreprocessor::checkKeyenceAgrees(const core::KeyenceData &keyence_data)
 {
-  bool keyence_data_disagrees = false;
+  KeyenceDataStatus current_data_status = KeyenceDataStatus::kDisagreed;
   for (std::size_t i = 0; i < keyence_data.size() - 1; ++i) {
-    if (keyence_data.at(i) != keyence_data.at(i + 1)) { keyence_data_disagrees = true; }
+    if (keyence_data.at(i) != keyence_data.at(i + 1)) {
+      current_data_status = KeyenceDataStatus::kDisagreed;
+    }
   }
 
-  if (keyence_data_disagrees && has_keyence_disagreed_) {
-    log_.log(hyped::core::LogLevel::kFatal, "Keyence disagreed more than once");
+  if (current_data_status == KeyenceDataStatus::kDisagreed
+      && previous_data_status_ == KeyenceDataStatus::kDisagreed) {
+    log_.log(core::LogLevel::kFatal, "Keyence disagreed more than once");
```

Comment:
We need to be careful with the logging here; if keyence gets out of sync once this will be spammed. Not sure how to solve this now, maybe add a `// TODOLater`?

Reply:
Could maintain a counter and check that counter while logging but i'll have a think about something similar to glog's `LOG_ONLY_ONCE` and `LOG_ONLY_N`

Reply:
Yeah that shouldn't be too hard to implement... though it would require some template/macro shenanigans if you want to keep it usable.

Reply:
I think the simplest solution is having `std::unordered_map<std::string, int> log_counts_;` in logger and 
```cpp
void Logger::logNTimes(const LogLevel level, int n, const char *format, ...)
{
  int &log_count = log_counts_.find(format)->second;
  if (log_count < n) {
    log(level, format);
    log_count++;
  }
}
```

Most standard logging libraries have something similar, I'll dig into some other definitions to see how they do it.